feat: add cancel button for context condensing operation

f14XuanLv opened the pull request:
### Related GitHub Issue

Closes: #6926 

### Roo Code Task Context (Optional)

### Description

This PR adds a cancel functionality for context condensing operations, addressing the issue where users couldn't interrupt long-running condensing processes.

Key implementation details:
- Added a cancel button (X icon) in the `TaskHeader` component that appears when `isCondensing` is true
- Implemented frontend-to-backend cancel message passing mechanism (`cancelCondenseContext` message type)
- Used `AbortController` API to gracefully cancel async stream operations
- Added abort signal checking in the `condenseContext` stream processing
- Properly distinguished between cancelled and failed operations in error handling
- Added internationalization support for English and Chinese

### Test Procedure

Testing steps:
1. Accumulate enough messages in a chat conversation to trigger context condensing
2. When condensing starts, observe the cancel button (X icon) appears in TaskHeader
3. Click the cancel button
4. Verify the condensing operation stops immediately
5. Check that the UI properly reflects the cancelled state (condensing indicator disappears)
6. Confirm no error messages are shown and the conversation can continue normally

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

[Screenshot 1: Shows cancel button during condensing]
<img width="556" height="750" alt="108270c3537d3904509e23d2c706295" src="https://github.com/user-attachments/assets/97453b23-d1eb-4ea9-a09c-365b87888f1c" />

[Screenshot 2: State after clicking cancel]
<img width="556" height="785" alt="ed2323c55e3395e1dc4abfc89fad57a" src="https://github.com/user-attachments/assets/1a09a907-34ba-4a2a-8907-79f3fb28da03" />

### Documentation Updates

- [x] No documentation updates are required.

### Additional Notes

This feature improves user experience, especially when dealing with large contexts. The cancellation operation uses the standard AbortController pattern, ensuring proper resource cleanup.

### Get in Touch

N/A - Available via GitHub @f14XuanLv

daniel-lxs commented:
Hey @f14XuanLv, I've been testing the cancel functionality and noticed a couple of issues:

1. The condensing cooldown seems to be broken - I can trigger condensing multiple times in a row even though there's supposed to be a limit that prevents condensing if one happened recently.

2. There's a weird bug with the cancel button - if I click condense, then cancel, then try to condense again, I get an error saying "Failed to condense, condense was cancelled" even though I'm trying to start a fresh condensing operation. Seems like the abort controller state might not be getting cleaned up properly?

I pushed a fix for the TypeScript type assertions to your branch, but these functional issues should probably be addressed before merging. The cancel state might need to be reset more thoroughly when starting a new condensing operation.

f14XuanLv commented:
Hey @daniel-lxs! 👋 

Thanks for the detailed feedback! I've addressed both issues you mentioned:

## 🔧 Fixed Issues

### 1. **Condensing cooldown mechanism** ✅
Added concurrency control check to prevent multiple condensing operations from running simultaneously:
```typescript
if (this.isCondensing) {
    console.warn("Context condensing is already in progress")
    return
}
```

### 2. **Cancel state cleanup** ✅  
Added intelligent cleanup waiting mechanism to avoid race conditions:
```typescript
if (this.condensingAbortController && this.condensingAbortController.signal.aborted) {
    await new Promise(resolve => setTimeout(resolve, 100))
}
```

## ✅ Testing Results

Verified through actual testing:
- **Cancel functionality**: Clicking cancel button immediately stops condensing operation ✅
- **Concurrency control**: Additional clicks during condensing are properly ignored ✅  
- **State cleanup**: After cancellation, there's ~5 second cooldown period, then condensing works normally ✅

Ready for re-review! 🚀

f14XuanLv commented:
Hi @daniel-lxs 

I need some clarification on a few points:

1. Version Reference Ambiguity:
When you say "there's some logic that prevents frequent condensing," which version are you referring to? Are you saying:
- The current version HAS this prevention logic, or 
- My PR HAS this prevention logic?

2. Screenshot Source:
Is the screenshot showing behavior from the current version or from my PR version?

3. Testing Methodology:
I'm curious about how you managed to trigger two condensing operations so quickly. In my testing, condensing operations typically take much longer than a few seconds to complete. Are you using a special testing environment, or did you encounter a specific scenario that allows rapid consecutive condensing? I haven't experienced cases where condensing completes within 5 seconds.

Understanding these details will help me better identify and fix the problem you've found.

Thanks for your patience!

f14XuanLv commented:
@hannesrudolph Could you please move this to dan's column so he can answer these questions?

---

Hi @daniel-lxs,

Thank you for the clarification. However, I'm confused about your statement regarding the existing cooldown logic on main.

## Current Analysis on main branch (commit: 87d50a78c)

After examining the codebase on the latest main branch (commit `87d50a78cb92ab5f050a4ec1233afe7004a52e3c`), I can only find **message-based prevention logic**, not time-based cooldown logic:

### 1. Message-based prevention in `src/core/condense/index.ts:119-124`:

```typescript
const keepMessages = messages.slice(-N_MESSAGES_TO_KEEP)  // Last 3 messages
const recentSummaryExists = keepMessages.some((message) => message.isSummary)
if (recentSummaryExists) {
    const error = t("common:errors.condensed_recently")
    return { ...response, error }
}
```

### 2. The time-related logic I found is in `Task.ts` for general API rate limiting:

```typescript
private static lastGlobalApiRequestTime?: number
// Used for general API rate limiting, not specifically for condensing
```

## My Questions:

1. Could you please point me to the **specific file and code location** where this time-based cooldown logic exists on main (commit `87d50a78c`)?
2. What is the **default cooldown window duration**?
3. Where is the **"last-completed-at timestamp"** being tracked that you mentioned?

I want to ensure I'm addressing the right issue, but I cannot locate the time-based cooldown mechanism you're referring to. It's possible I'm missing something, so your guidance on the specific location would be very helpful.

## Regarding the reproduction steps:

The behavior you described (immediate re-triggering after completion/cancellation) might be related to the **concurrent execution guard** (`isCondensing`) rather than a time-based cooldown, which could explain why my changes affect it.

Could you help me identify where this time-based cooldown logic is implemented on main? This will help me preserve it properly in my PR.

Thanks for your patience!
